### PR TITLE
Close channels on my turn

### PR DESCRIPTION
**Pull request by andrewgordstewart**
Fixes https://github.com/statechannels/monorepo/issues/1811

# Closing channel changes
- Refactors `ChannelState` to have `payer` and `beneficiary` properties of type `Peer`. I think this makes it more understandable and maintainable
- Adds a `canUpdateChannel` method to the `PaymentChannelClient`
   * this makes `getLatestPaymentReceipt` redundant. If you are making a payment, you are the payer. If you are the payer, and it's your turn, then you have the latest payment receipt
   * I think having one (hopefully correct 😬 ) method that tells you "You can update the channel" simplifies things
- Refactors some payment channel client code
   * onChannelUpdated/Proposed seemed like it was needlessly complicated
   * `PaymentChannelClient` exposes a `channelState` observable which converts a `ChannelResult` to a `ChannelState`, for a specific channel. The rest of the code does not need to observe `ChannelClient.channelState`, and then filter on `channelId`

# Other changes
- [xstate-wallet] fixes error codes (also fixed inhttps://github.com/statechannels/monorepo/pull/1987 )
- [client-api-schema] exports `ErrorCodes`

# Current status
1. the stress test and cancelling test passes somewhat frequently, locally.
2. cancelling _always_ passed locally, in headful mode. I think it creates fortuitous timing
3. cancelling and closing does not necessarily work reliably: https://github.com/statechannels/monorepo/issues/1811#issuecomment-635029656 showed me that this sequence of events can happen:
   1. W3T learns it's my turn, and call `UpdateChannel` as a part of `makePayment`
   2. W3T decides to cancel the torrent, learn it's my turn, and call `CloseChannel`
   3. Wallet receives `UpdateChannel`, and updates the channel
   4. Wallet receives `CloseChannel`, it's not my turn, and returns an error

Rather than dealing with 3 right now, I am opting to merge this stuff right away, as it seems like an improvement and is getting big.

### CODE REVIEW

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 1]
```diff
@@ -26,24 +27,101 @@ function sanitizeMessageForFirebase(message) {
...
+    status
+  } = channelResult;
+
+  return {
+    channelId,
+    turnNum: utils.bigNumberify(turnNum),
+    status,
+    challengeExpirationTime,
+    beneficiary: {
+      signingAddress: participants[0].participantId,
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:75]: Absolutely!

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 2]
```diff
@@ -205,15 +269,25 @@ export class PaymentChannelClient {
...
-    if (!['closing', 'closed'].includes(this.channelCache[channelId].status)) {
-      await this.getLatestPaymentReceipt(channelId);
-      this.channelClient.closeChannel(channelId);
-    }
-    return this.channelClient.channelState
-      .pipe(map(convertToChannelState), filter(isClosed), first())
+    return this.channelState(channelId)
+      .pipe(
+        first(cs => cs.status === 'closed'),
+        tap(() => closing.unsubscribe())
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:289]: This is to protect when there are zero events that get past the `first` operator (which seems possible)

File: packages/web3torrent/src/clients/payment-channel-client.ts [hunk 3]
```diff
@@ -205,15 +269,25 @@ export class PaymentChannelClient {
   }
 
   async closeChannel(channelId: string): Promise<ChannelState> {
-    const isClosed = (channelState: ChannelState) =>
-      channelState.channelId === channelId && channelState.status === 'closed';
+    logger.info(`Waiting for my turn to close channel ${channelId}`);
+    // Let an existing channel update happen before waiting for my turn
+    await new Promise(resolve => setTimeout(resolve, 500));
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:274]: > Maybe leave a comment explaining the delay

I think the existing comment, which was meant to explain the delay, might not be clear enough 😆

File: packages/web3torrent/src/clients/payment-channel-client.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/web3torrent/src/clients/payment-channel-client.ts:302]: Yep, that makes sense.

File: packages/web3torrent/src/clients/payment-channel-client.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/web3torrent/src/clients/payment-channel-client.ts:282]: https://github.com/statechannels/monorepo/pull/1997/commits/023e7abcf8c8adfb14f7fd53de3e0f287cfea525

Does that make it clear?

**Pull request merged**